notebooks/week2/05.log_and_register_fe_model.py

```diff
@@ -0,0 +1,193 @@
+)
+
+# COMMAND ----------
+
+# Load feature-engineered DataFrame
+model_feature_lookups = [FeatureLookup(table_name=feature_table_name, lookup_key=id)]
+training_set = fe.create_training_set(
+    df=train_set[[id, target]], feature_lookups=model_feature_lookups, label=target, exclude_columns=[id]
+)
+
+training_df = training_set.load_df().toPandas()
+training_df.info()
+
+# COMMAND ----------
+
+# Split features and target
+X_train = training_df[num_features + cat_features]
+y_train = training_df[target]
+X_test = test_set[num_features + cat_features]
+y_test = test_set[target]
+
+# COMMAND ----------
+
+# Create preprocessing steps for numeric and categorical data
+numeric_transformer = Pipeline(steps=[("imputer", SimpleImputer(strategy="median")), ("scaler", StandardScaler())])
+
+categorical_transformer = Pipeline(
+    steps=[
+        ("imputer", SimpleImputer(strategy="constant", fill_value="missing")),
+        ("onehot", OneHotEncoder(handle_unknown="ignore")),
+    ]
+)
+
+# Combine preprocessing steps
+preprocessor = ColumnTransformer(
+    transformers=[
+        ("num", numeric_transformer, num_features),
+        ("cat", categorical_transformer, cat_features),
+    ]
+)
+
+# COMMAND ----------
+
+model = Pipeline(steps=[("preprocessor", preprocessor), ("classifier", RandomForestClassifier(**parameters))])
+
+# COMMAND ----------
+
+# Set and start MLflow experiment
+mlflow.set_experiment(experiment_name="/Shared/hotel-reservation-fe")
+git_sha = "3055af355f360ba5784ae7037f3260a70331f702"
+
+with mlflow.start_run(tags={"branch": "week2", "git_sha": f"{git_sha}"}) as run:
+    run_id = run.info.run_id
+    model.fit(X_train, y_train)
+    y_pred = model.predict(X_test)
+
+    # Calculate and print metrics
+    f1 = f1_score(y_test, y_pred, average="weighted")
+    cm = confusion_matrix(y_test, y_pred)
+    cm_display = ConfusionMatrixDisplay(confusion_matrix=cm)
+    cm_display.plot()
+    plt.savefig("confusion-matrix.png")
+    plt.show()
+
+    print(f"F1 Score: {f1}")
+    print(f"Confusion Matrix: {cm}")
+
+    # Log model parameters, metrics, and model
+    mlflow.log_param("model_type", "RandomForestClassifier")
+    mlflow.log_params(parameters)
+    mlflow.log_metric("f1 score", f1)
+    mlflow.log_artifact("confusion-matrix.png")
+    signature = infer_signature(model_input=X_train, model_output=y_pred)
+
+    # Log model with feature engineering
+    fe.log_model(
+        model=model,
+        flavor=mlflow.sklearn,
+        artifact_path="RandomForest",
+        training_set=training_set,
+        signature=signature,
+    )
+
+mlflow.register_model(
+    model_uri=f"runs:/{run_id}/RandomForestClassifier-fe",
```

Comment:
artifact path and where the model is logged are not the same - this code would fail.